aptos-move/framework/aptos-framework/sources/account.move

```diff
@@ -326,6 +341,113 @@ module aptos_framework::account {
+
+        // update the `OriginatingAddress` table, so we can find the originating address using the new address
+        let new_address = from_bcs::to_address(new_auth_key);
+        let address_map = &mut borrow_global_mut<OriginatingAddress>(@aptos_framework).address_map;
+        if (table::contains(address_map, curr_auth_key)) {
+            assert!(rotation_cap_offerer_address == table::remove(address_map, curr_auth_key), error::not_found(EINVALID_ORIGINATING_ADDRESS));
+        };
+        table::add(address_map, new_address, rotation_cap_offerer_address);
+
+        // update the authentication key of the offerer's account
+        let offerer_account_resource = borrow_global_mut<Account>(rotation_cap_offerer_address);
+        event::emit_event<KeyRotationEvent>(
+            &mut offerer_account_resource.key_rotation_events,
+            KeyRotationEvent {
+                old_authentication_key: offerer_account_resource.authentication_key,
+                new_authentication_key: new_auth_key,
+            }
+        );
+
+        offerer_account_resource.authentication_key = new_auth_key;
+    }
+
+    /// Offers rotation capability on behalf of `account` to the account at address `recipient_address`.
+    /// An account can delegate its rotation capability to only one other address at one time.
+    /// `rotation_capability_sig_bytes` is the `RotationCapabilityOfferProofChallengeV2` signed by the account owner's key.
+    /// `account_scheme` is the scheme of the account (ed25519 or multi_ed25519).
+    /// `account_public_key_bytes` is the public key of the account owner.
+    /// `recipient_address` is the address of the recipient of the rotation capability - note that if there's an existing
+    /// `recipient_address` in the account owner's `RotationCapabilityOffer`, this will replace the
+    /// previous `recipient_address` upon successful verification (the previous recipient will no longer have access
+    /// to the account owner's rotation capability).
+    public entry fun offer_rotation_capability(
+        account: &signer,
+        rotation_capability_sig_bytes: vector<u8>,
+        account_scheme: u8,
+        account_public_key_bytes: vector<u8>,
+        recipient_address: address,
+    ) acquires Account {
+        let addr = signer::address_of(account);
+        assert!(exists_at(recipient_address), error::not_found(EACCOUNT_DOES_NOT_EXIST));
+
+        // proof that this account intends to delegate its rotation capability to another account
+        let account_resource = borrow_global_mut<Account>(addr);
+        let proof_challenge = RotationCapabilityOfferProofChallengeV2 {
+            chain_id: chain_id::get(),
+            sequence_number: account_resource.sequence_number,
+            source_address: addr,
+            recipient_address,
+        };
+
+        // verify the signature on `RotationCapabilityOfferProofChallengeV2` by the account owner
+        if (account_scheme == ED25519_SCHEME) {
```

Comment:
Can you add documentation on why we need the challenge again? The signer for the account that wants to offer the rotation capability is present so that means they explicitly signed this transaction. It seems a bit redundant to require an extra signed message here.

Reply:
@davidiw can confirm: this is to prevent malicious contracts, who would also have such a signer for the account interacting with the contract, to call into this function and maliciously rotate the key.

While technically speaking, users should make sure they trust the contract they are interacting with, in practice, users will rely on what the wallet displays to them as the transaction's outcome.

But due to a classic "time-of-check time-of-use" attack, what the wallet displays to the user, is not necessarily what the contract will do. For example, consider a contract that with 50% probability (based on the current timestamp in Move), rotates somebody's key. The wallet might be unlucky and get an outcome where nothing is rotated, incorrectly telling the user nothing bad will happen. But when the transaction actually gets executed, the attacker gets lucky and the execution path triggers the account key rotation.

We prevent such attacks by asking for this extra signature authorizing the key rotation.

Reply:
Hmm I see that makes sense. We also have a similar safeguard in rotate_authentication_key so it's all consistent.

Reply:
Would still be great to document this in the function's docs, similar to rotate_authentication_key's docs.

Reply:
@0xchloe

Reply:
okay will add more comments : ) thanks @alinush for explaining